RDS: The index is out of array

Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

### Terraform Version
Run `terraform -v` to show the version. If you are not running the latest version of Terraform, please upgrade because your issue may have already been fixed.


Terraform v0.12.3
+ provider.opentelekomcloud v1.11.0

### Affected Resource(s)
Please list the resources as a list, for example:
- opentelekomcloud_rds_instance_v3

If this issue appears to affect multiple resources, it may be an issue with Terraform's core, so please mention this.

### Terraform Configuration Files
```hcl

data "opentelekomcloud_networking_secgroup_v2" "db_sec_group" {
  name = "database-security"
}

resource "opentelekomcloud_rds_instance_v3" "sql_instance2" {

  name              = "rds-instance-${var.area}-${var.environment}-${var.count}"
  availability_zone = ["eu-de-02"]
  flavor            = "rds.mssql.s1.xlarge"
  security_group_id = data.opentelekomcloud_networking_secgroup_v2.db_sec_group.id
  subnet_id         = var.netSubId
  vpc_id            = var.netVPCId

  db {
    password = "supersecretpassword"
    type     = "SQLServer"
    version  = "2016 EE"
    port     = "1433"
  }

  volume {
    size               = 500
    type               = "COMMON"
    disk_encryption_id = var.vmKeyID
  }
}
```

### Debug Output

2019/08/16 13:20:37 [DEBUG] module.Grid.module.grid-server.opentelekomcloud_rds_instance_v3.sql_instance2: apply errored, but we're indicating that via the Error pointer rather than returning it: Error building the request body of api(create), err=navigate:: The index is out of array
2019/08/16 13:20:37 [TRACE] module.Grid.module.grid-server: eval: *terraform.EvalMaybeTainted
2019/08/16 13:20:37 [ERROR] module.Grid.module.grid-server: eval: *terraform.EvalApplyPost, err: Error building the request body of api(create), err=navigate:: The index is out of array
2019/08/16 13:20:37 [ERROR] module.Grid.module.grid-server: eval: *terraform.EvalSequence, err: Error building the request body of api(create), err=navigate:: The index is out of array
20

### Panic Output
If Terraform produced a panic, please provide a link to a GitHub Gist containing the output of the `crash.log`.

### Expected Behavior
What should have happened?

RDS is created

### Actual Behavior
What actually happened?

stops with the error:

Error: Error building the request body of api(create), err=navigate:: The index is out of array

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform apply`

--- a/opentelekomcloud/transport.go
+++ b/opentelekomcloud/transport.go
@@ -119,25 +119,38 @@ func navigateMap(d interface{}, index []string) (interface{}, error) {
 
 func navigateValue(d interface{}, index []string, arrayIndex map[string]int) (interface{}, error) {
 	for n, i := range index {
+		if d == nil {
+			return nil, nil
+		}
 		if d1, ok := d.(map[string]interface{}); ok {
 			d, ok = d1[i]
 			if !ok {
-				return nil, fmt.Errorf("navigate:: '%s' may not exist", i)
+				msg := fmt.Sprintf("navigate value with index(%s)", strings.Join(index, "."))
+				return nil, fmt.Errorf("%s: '%s' may not exist", msg, i)
 			}
 		} else {
-			return nil, fmt.Errorf("navigateValue:: Can not convert (%s) to map, index=%s", reflect.TypeOf(d), strings.Join(index, "."))
+			msg := fmt.Sprintf("navigate value with index(%s)", strings.Join(index, "."))
+			return nil, fmt.Errorf("%s: Can not convert (%s) to map", msg, reflect.TypeOf(d))
 		}
 
 		if arrayIndex != nil {
 			if j, ok := arrayIndex[strings.Join(index[:n+1], ".")]; ok {
+				if d == nil {
+					return nil, nil
+				}
 				if d2, ok := d.([]interface{}); ok {
+					if len(d2) == 0 {
+						return nil, nil
+					}
 					if j >= len(d2) {
-						return nil, fmt.Errorf("navigate:: The index is out of array")
+						msg := fmt.Sprintf("navigate value with index(%s)", strings.Join(index, "."))
+						return nil, fmt.Errorf("%s: The index is out of array", msg)
 					}
 
 					d = d2[j]
 				} else {
-					return nil, fmt.Errorf("navigateValue:: Can not convert (%s) to array, index=%s.%v", reflect.TypeOf(d), i, j)
+					msg := fmt.Sprintf("navigate value with index(%s)", strings.Join(index, "."))
+					return nil, fmt.Errorf("%s: Can not convert (%s) to array, index=%s.%v", msg, reflect.TypeOf(d), i, j)
 				}
 			}
 		}

